diff --git a/binding.gyp b/binding.gyp
index 68a8f58..e24e5d0 100644
--- a/binding.gyp
+++ b/binding.gyp
@@ -82,7 +82,6 @@
             ['OS == "mac"', {
               'libraries': [
                 '<(module_root_dir)/build/libzmq/lib/libzmq.a',
-                "<!@(pkg-config libsodium --libs)",
               ],
             }],

@@ -155,7 +154,6 @@
                       # 1 - MultiThreadedDebug (/MTd)
                       # 2 - MultiThreadedDLL (/MD)
                       # 3 - MultiThreadedDebugDLL (/MDd)
-                      'RuntimeLibrary': 3,
                       'AdditionalOptions': [
                         '-std:c++17',
                         "/DEBUG",
@@ -172,7 +170,6 @@
                     # # C:/Program Files (x86)/Microsoft Visual Studio/2019/Preview/VC/Tools/MSVC/14.29.29917/bin/Hostx64/x64/
                     'VCCLCompilerTool': {
                       'ExceptionHandling': 2,       # /EHsc
-                      'RuntimeLibrary': 3,
                       "DebugInformationFormat": "ProgramDatabase", # /Zi
                       'AdditionalOptions': [
                         '-std:c++17',
diff --git a/package.json b/package.json
index 058ecd1..12fcd8a 100644
--- a/package.json
+++ b/package.json
@@ -20,7 +20,7 @@
   "dependencies": {
     "@aminya/node-gyp-build": "4.5.0-aminya.4",
     "cross-env": "^7.0.3",
-    "node-addon-api": "^5.0.0",
+    "node-addon-api": "^6.0.0",
     "shelljs": "^0.8.5",
     "shx": "^0.3.4"
   },
diff --git a/script/build.ts b/script/build.ts
index c055ebc..717a236 100644
--- a/script/build.ts
+++ b/script/build.ts
@@ -27,11 +27,11 @@ function main() {

   // https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html
   if (process.platform === "win32") {
-    if (CMAKE_BUILD_TYPE !== "Debug") {
-      build_options += " -DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreadedDLL"
-    } else {
-      build_options += " -DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreadedDebugDLL"
-    }
+      if (CMAKE_BUILD_TYPE !== "Debug") {
+        build_options += " -DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded"
+      } else {
+        build_options += " -DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreadedDebug"
+      }
   }

   build_options += archCMakeOptions()
@@ -78,7 +78,7 @@ function main() {
     writeFileSync(clang_format_file, "")
   }

-  const cmake_configure = `cmake -S "${src_dir}" -B ./build ${build_options} -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE} -DCMAKE_INSTALL_PREFIX="${libzmq_install_prefix}" -DCMAKE_INSTALL_LIBDIR=lib -DBUILD_STATIC=ON -DBUILD_TESTS=OFF -DBUILD_SHARED=OFF -DWITH_DOCS=OFF -DWITH_LIBSODIUM=ON -DWITH_LIBSODIUM_STATIC=ON`
+  const cmake_configure = `cmake -S "${src_dir}" -B ./build ${build_options} -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE} -DCMAKE_INSTALL_PREFIX="${libzmq_install_prefix}" -DCMAKE_INSTALL_LIBDIR=lib -DBUILD_STATIC=ON -DBUILD_TESTS=OFF -DBUILD_SHARED=OFF -DWITH_DOCS=OFF -DWITH_LIBSODIUM=OFF`
   console.log(cmake_configure)
   exec(cmake_configure, execOptions)

